posts/inside-rust/2023-11-09-spec-vision.md

```diff
@@ -0,0 +1,206 @@
+
+Our answer:
+That stability promise left a bit of wiggle-room for itself, in terms of what the project considers "fear" vs "reasonable labor associated with a Rust upgrade."
+When defining semantics, one must be more explicit about any such wiggle-room.
+Some Rust users *need* a description of the expected semantics as it stands for the Rust release that sits in their hands; they are the audience for version-specific details.
+But other Rust users, such as some library developers, have a more forward-looking perspective.
+The forward-looking developers may require an assurance that one specific code snippet A will always be accepted, and will also always have a particular meaning.
+They may require an assurance that a different snippet B will *never* be accepted.
+Or they may require an assurance that a third snippet C leveraging Unsafe Rust will always have undefined behavior (e.g. to justify a local transformation by arguing that no *new* undefined behavior is injected by that transformation.)
+These are all cases that call for a prescriptive definition of the semantics; it does not matter in these cases what the compiler currently does -- what matters is what it will do in the future, which is inherently prescriptive.
+
+It would be premature to fix firm definitions in those areas, e.g. categorizing for each input program whether it is accepted or rejected by the type inference system, and then forcing all futures versions of Rust to follow that same categorization.
+Another similar example: If we chose a fixed grammar, and then said all future versions of Rust must strictly categorize all source inputs as accepted or rejected according to that one grammar, then that would restrict our ability to add future backward-compatible language extensions to the grammar.
+Therefore, these kinds of guarantees (especially with respect to details of the type inference rules, or details of what unsafe code is well-defined) are where the prescriptive *bounds* arise.
+Such bounds allow for a middle ground of programs, where we do not commit all future versions of Rust to always make the same decision that the current version makes.
+For example, one can then say, prescriptively, that a given grammar provides a lower bound on the set of programs that must be accepted by all future versions of Rust, while still allowing the language to evolve in a backward compatible fashion.
+One can also say, descriptively, that the current version of Rust rejects source inputs that do not conform to the grammar.
+
+The descriptive definitions tell the reader how a construct will behave with respect to that Rust version; the prescriptive bounds tell the reader what they can and cannot expect to hold true in the future of Rust.
+
+Thus, we conclude that an ideal specification will need to address both the static and dynamic semantics, for both the prescriptive bounds and descriptive details.
```

Comment:
```suggestion
Thus, we conclude that an ideal specification will need to use both prescriptive bounds and descriptive details, for both the static and dynamic semantics.
```

or

```suggestion
Thus, we conclude that an ideal specification will need to address both the static and dynamic semantics, with both prescriptive bounds and descriptive details.
```

Reply:
I like your first suggestion, though I think I would opt to say the specification will "provide" both prescriptive bounds and descriptive details, rather than "use"